File: Day80/Part5/weapon_path.py
import sys
import heapq

def solve():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return

    iterator = iter(data)
    try:
        N = int(next(iterator))
        M = int(next(iterator))
        S = int(next(iterator))
        E = int(next(iterator))
        InitDur = int(next(iterator))
        
        adj = [[] for _ in range(N + 1)]
        for _ in range(M):
            u = int(next(iterator))
            v = int(next(iterator))
            w = int(next(iterator))
            adj[u].append((v, w))
            adj[v].append((u, w))
    except StopIteration:
        return

    # Max heap: store (-durability, u, repaired)
    pq = [(-InitDur, S, 0)]
    
    # max_dur[u][repaired]
    max_dur = [[-1] * 2 for _ in range(N + 1)]
    max_dur[S][0] = InitDur
    
    while pq:
        d_neg, u, r = heapq.heappop(pq)
        d = -d_neg
        
        if d < max_dur[u][r]:
            continue
        
        if u == E:
            print("true")
            return
        
        # Option 1: Move
        for v, w in adj[u]:
            new_d = d - w
            if new_d >= 0:
                if new_d > max_dur[v][r]:
                    max_dur[v][r] = new_d
                    heapq.heappush(pq, (-new_d, v, r))
                    
        # Option 2: Repair
        if r == 0:
            new_d = InitDur
            if new_d > max_dur[u][1]:
                max_dur[u][1] = new_d
                heapq.heappush(pq, (-new_d, u, 1))
                
    print("false")

if __name__ == "__main__":
    solve()

File: Day80/Part5/weapon_path.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>

using namespace std;

struct Edge {
    int to;
    int weight;
};

struct State {
    int durability;
    int u;
    int repaired; // 0 or 1

    bool operator<(const State& other) const {
        return durability < other.durability; // Max heap
    }
};

int main() {
    int N, M, S, E, InitDur;
    if (!(cin >> N >> M >> S >> E >> InitDur)) return 0;

    vector<vector<Edge>> adj(N + 1);
    for (int i = 0; i < M; ++i) {
        int u, v, w;
        cin >> u >> v >> w;
        adj[u].push_back({v, w});
        adj[v].push_back({u, w}); // Assuming undirected? "Graph berbobot". Usually undirected unless specified.
        // "Path" usually implies directed or undirected. Let's assume undirected for standard map.
    }

    // dist[u][repaired] = max durability
    vector<vector<int>> max_dur(N + 1, vector<int>(2, -1));
    priority_queue<State> pq;

    max_dur[S][0] = InitDur;
    pq.push({InitDur, S, 0});

    while (!pq.empty()) {
        State curr = pq.top();
        pq.pop();

        int d = curr.durability;
        int u = curr.u;
        int r = curr.repaired;

        if (d < max_dur[u][r]) continue;
        if (u == E) {
            cout << "true" << endl;
            return 0;
        }

        // Option 1: Move to neighbor
        for (auto& edge : adj[u]) {
            int v = edge.to;
            int w = edge.weight;
            int new_d = d - w;

            if (new_d >= 0) {
                if (new_d > max_dur[v][r]) {
                    max_dur[v][r] = new_d;
                    pq.push({new_d, v, r});
                }
            }
        }

        // Option 2: Repair (if not yet repaired)
        if (r == 0) {
            // Repair resets durability to InitDur
            // We stay at u, but state changes to 1
            int new_d = InitDur;
            if (new_d > max_dur[u][1]) {
                max_dur[u][1] = new_d;
                pq.push({new_d, u, 1});
            }
        }
    }

    cout << "false" << endl;

    return 0;
}
